conn.go

```diff
@@ -118,12 +118,12 @@ func (c *Conn) Streams() []*Stream {
 // Close closes this connection
 func (c *Conn) Close() error {
 	c.closeLock.Lock()
-	defer c.closeLock.Unlock()
-
 	if c.closed {
+		c.closeLock.Unlock()
```

Comment:
this wont work because then a second caller could exit way before the Conn is _actually closed_. maybe make them wait?

```go
c.closeLock.Lock()
if c.closed {
  c.closeLock.Unlock()
  <-c.closeChan
  return c.closeErr
}
c.closed = true
c.closeLock.Unlock()
defer func() {
  c.closeChan<- struct{}{}
}()
```

or use https://github.com/jbenet/goprocess and avoid the close lock madness alltogether, which also gives us nice errors. (goproc doesnt have the UX bugs, it's goproc/ratelimit that has caused the problems in the last few months)